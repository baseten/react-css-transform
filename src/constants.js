import PropTypes from 'prop-types';
import { glMatrix } from 'gl-matrix';

export const MULTIPLICATION_ORDER = {
  /**
   * This is the default, more natural mathematically and (arguably) more useful, it behaves in the same order as when
   * applying multiple transforms on the css `transform` property:
   *
   * <Transform2d translate={{x:100, y:100}} multiplicationOrder={Transform2d.MULTIPLICATION_ORDER.POST}>
   *   <Transform2d rotate={0.7854}>
   *     <Transform2d scale={2}>
   *       <div></div>
   *     </Transform2d>
   *   </Transform>
   * </Transform2d>
   *
   * Is the same as applying a style of `transform: translate(100px,100px) rotate(0.7845rad) scale(2)` to the inner div
   */
  POST: 'POST',
  /**
   * This mimics how CSS transforms work between nested DOM elements
   *
   * <Transform2d translate={{x:100, y:100}} multiplicationOrder={Transform2d.MULTIPLICATION_ORDER.PRE}>
   *   <Transform2d rotate={0.7854}>
   *     <Transform2d scale={2}>
   *       <div></div>
   *     </Transform2d>
   *   </Transform>
   * </Transform2d>
   *
   * Is basically the same (assuming all the nested divs are positioned absolutely in the same place) as:
   *
   * <div style="transform: translate(100px, 100px)>
   *   <div style="transform: rotate(0.7854rad)">
   *     <div style="transform: scale(2)">
   *       <div></div>
   *     </div>
   *   </div>
   * </div>
   */
  PRE: 'PRE',
};

export const vec2Shape = PropTypes.shape({
  x: PropTypes.number.isRequired,
  y: PropTypes.number.isRequired,
});

export const vec3Shape = PropTypes.shape({
  x: PropTypes.number.isRequired,
  y: PropTypes.number.isRequired,
  z: PropTypes.number.isRequired,
});

export const glMatrixType = PropTypes.instanceOf(glMatrix.ARRAY_TYPE);
