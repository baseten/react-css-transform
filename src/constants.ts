import {
  isVec2,
  isVec3,
  isMat2d,
  isMat4,
  isVec2Object,
  isVec3Object,
} from './utils';

type PropTypes = { [key: string]: any };

export const vec2Obj = <P extends PropTypes>(
  props: P,
  propName: string & keyof P,
  componentName: string,
) => {
  const passes = isVec2Object(props[propName]);
  return passes
    ? null
    : new Error(`${propName} in ${componentName} is not a vec2 shape`);
};

export const vec3Obj = <P extends PropTypes>(
  props: P,
  propName: string & keyof P,
  componentName: string,
) => {
  const passes = isVec3Object(props[propName]);
  return passes
    ? null
    : new Error(`${propName} in ${componentName} is not a vec3 shape`);
};

export const vec2GlMatrix = <P extends PropTypes>(
  props: P,
  propName: string & keyof P,
  componentName: string,
) => {
  const passes = isVec2(props[propName]);
  return passes
    ? null
    : new Error(`${propName} in ${componentName} is not a gl-matrix vec2`);
};

export const vec3GlMatrix = <P extends PropTypes>(
  props: P,
  propName: string & keyof P,
  componentName: string,
) => {
  const passes = isVec3(props[propName]);
  return passes
    ? null
    : new Error(`${propName} in ${componentName} is not a gl-matrix vec3`);
};

export const mat2dGlMatrix = <P extends PropTypes>(
  props: P,
  propName: string & keyof P,
  componentName: string,
) => {
  const passes = isMat2d(props[propName]);
  return passes
    ? null
    : new Error(`${propName} in ${componentName} is not a gl-matrix mat2d`);
};

export const mat4GlMatrix = <P extends PropTypes>(
  props: P,
  propName: string & keyof P,
  componentName: string,
) => {
  const passes = isMat4(props[propName]);
  return passes
    ? null
    : new Error(`${propName} in ${componentName} is not a gl-matrix mat4`);
};
