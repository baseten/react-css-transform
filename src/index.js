export {
  MULTIPLICATION_ORDER,
  vec2Obj,
  vec3Obj,
  vec2GlMatrix,
  vec3GlMatrix,
  mat2dGlMatrix,
  mat4GlMatrix,
} from './constants';
export { default as Transform2d } from './Transform2d';
export { default as Transform3d } from './Transform3d';
