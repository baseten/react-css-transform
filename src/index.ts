export * from './propTypes';
export { useFactoryRef } from './useFactoryRef';
export {
  get2dCSSMatrixString,
  get3dCSSMatrixString,
  getCSSMatrixString,
  merge2dMatrixWithStyles,
  merge3dMatrixWithStyles,
} from './utils';
export { default as Transform2d, use2dTransformations } from './Transform2d';
export { default as Transform3d, use3dTransformations } from './Transform3d';
export type { Transform2dProps } from './Transform2d';
export type { Transform3dProps } from './Transform3d';
export type { GLMatrixType, Vec2Object, Vec3Object } from './types';
