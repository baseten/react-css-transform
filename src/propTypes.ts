import {
  isNil,
  isVec2,
  isVec3,
  isMat2d,
  isMat4,
  isVec2Object,
  isVec3Object,
} from './utils';

type PropTypes = { [key: string]: any };
type PropTypeValidator = {
  <P extends PropTypes>(
    props: P,
    propName: string & keyof P,
    componentName: string,
  ): null | Error;
};
type PropTypeWithRequired = PropTypeValidator & {
  isRequired: PropTypeValidator;
};

function createPropType<P extends PropTypes>(
  validator: PropTypeValidator,
  isRequired: boolean,
) {
  return (props: P, propName: string & keyof P, componentName: string) => {
    const prop = props[propName];

    if (isNil(prop)) {
      if (isRequired) {
        return new Error(
          `${propName} is required in ${componentName} but received ${prop}`,
        );
      }
      return null;
    }

    return validator(props, propName, componentName);
  };
}

const vec2ObjValidator = <P extends PropTypes>(
  props: P,
  propName: string & keyof P,
  componentName: string,
) => {
  const passes = isVec2Object(props[propName]);
  return passes
    ? null
    : new Error(`${propName} in ${componentName} is not a vec2 shape`);
};

export const vec2Obj: PropTypeWithRequired = Object.assign(
  createPropType(vec2ObjValidator, false),
  {
    isRequired: createPropType(vec2ObjValidator, true),
  },
);

const vec3ObjValidator = <P extends PropTypes>(
  props: P,
  propName: string & keyof P,
  componentName: string,
) => {
  const passes = isVec3Object(props[propName]);
  return passes
    ? null
    : new Error(`${propName} in ${componentName} is not a vec3 shape`);
};

export const vec3Obj: PropTypeWithRequired = Object.assign(
  createPropType(vec3ObjValidator, false),
  { isRequired: createPropType(vec3ObjValidator, true) },
);

const vec2GlMatrixValidator = <P extends PropTypes>(
  props: P,
  propName: string & keyof P,
  componentName: string,
) => {
  const passes = isVec2(props[propName]);
  return passes
    ? null
    : new Error(`${propName} in ${componentName} is not a gl-matrix vec2`);
};

export const vec2GlMatrix: PropTypeWithRequired = Object.assign(
  createPropType(vec2GlMatrixValidator, false),
  {
    isRequired: createPropType(vec2GlMatrixValidator, true),
  },
);

const vec3GlMatrixValidator = <P extends PropTypes>(
  props: P,
  propName: string & keyof P,
  componentName: string,
) => {
  const passes = isVec3(props[propName]);
  return passes
    ? null
    : new Error(`${propName} in ${componentName} is not a gl-matrix vec3`);
};
export const vec3GlMatrix: PropTypeWithRequired = Object.assign(
  createPropType(vec3GlMatrixValidator, false),
  {
    isRequired: createPropType(vec3GlMatrixValidator, true),
  },
);

const mat2dGlMatrixValidator = <P extends PropTypes>(
  props: P,
  propName: string & keyof P,
  componentName: string,
) => {
  const passes = isMat2d(props[propName]);
  return passes
    ? null
    : new Error(`${propName} in ${componentName} is not a gl-matrix mat2d`);
};

export const mat2dGlMatrix: PropTypeWithRequired = Object.assign(
  createPropType(mat2dGlMatrixValidator, false),
  { isRequired: createPropType(mat2dGlMatrixValidator, true) },
);

export const mat4GlMatrixValidator = <P extends PropTypes>(
  props: P,
  propName: string & keyof P,
  componentName: string,
) => {
  const passes = isMat4(props[propName]);
  return passes
    ? null
    : new Error(`${propName} in ${componentName} is not a gl-matrix mat4`);
};

export const mat4GlMatrix: PropTypeWithRequired = Object.assign(
  createPropType(mat4GlMatrixValidator, false),
  { isRequired: createPropType(mat4GlMatrixValidator, true) },
);
