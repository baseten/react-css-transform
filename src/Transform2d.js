import React from 'react';
import PropTypes from 'prop-types';
import { mat2d, vec2 } from 'gl-matrix';

import { MULTIPLICATION_ORDER, vec2Shape, glMatrixType } from './constants';

export default class Transform2d extends React.Component {
  static propTypes = {
    parentMatrixWorld: glMatrixType,
    multiplicationOrder: PropTypes.oneOf([MULTIPLICATION_ORDER.PRE, MULTIPLICATION_ORDER.POST]),
    translate: vec2Shape,
    scale: PropTypes.oneOfType([vec2Shape, glMatrixType, PropTypes.number]),
    rotate: PropTypes.number,
    children: PropTypes.node.isRequired,
  };

  static defaultProps = {
    parentMatrixWorld: mat2d.create(),
    multiplicationOrder: MULTIPLICATION_ORDER.POST,
  };

  matrix = mat2d.create();
  matrixWorld = mat2d.create();
  vTranslation = vec2.create();
  vScale = vec2.create();

  render() {
    const { children, parentMatrixWorld, multiplicationOrder, translate, scale, rotate } = this.props;
    const { matrix, matrixWorld, vTranslation, vScale } = this;

    const tx = translate && translate.x ? translate.x : 0;
    const ty = translate && translate.y ? translate.y : 0;

    let sx, sy;

    if (typeof scale === 'number') {
      sx = scale;
      sy = scale;
    } else {
      sx = scale && scale.x ? scale.x : 1;
      sy = scale && scale.y ? scale.y : 1;
    }

    vec2.set(vTranslation, tx, ty);
    vec2.set(vScale, sx, sy);

    mat2d.identity(matrix);

    // T * R * S
    mat2d.translate(matrix, matrix, vTranslation);
    mat2d.rotate(matrix, matrix, rotate || 0);
    mat2d.scale(matrix, matrix, vScale);

    if (multiplicationOrder === MULTIPLICATION_ORDER.PRE) {
      mat2d.multiply(matrixWorld, matrix, parentMatrixWorld);
    } else if (multiplicationOrder === MULTIPLICATION_ORDER.POST) {
      mat2d.multiply(matrixWorld, parentMatrixWorld, matrix);
    }

    if (typeof children === 'function') {
      return children({
        parentMatrixWorld: matrixWorld,
        multiplicationOrder,
        style: {
          transform: matrixWorld.join(','),
        },
      });
    }

    return React.Children.map(children, this.renderChild);
  }

  renderChild = child => {
    const { multiplicationOrder } = this.props;
    const { matrixWorld } = this;

    const childStyle = child.props.style || {};
    const style = {
      ...childStyle,
      transform: `matrix(${matrixWorld.join(',')})`,
    };

    const childProps = {
      style,
    };

    if (typeof child.type !== 'string') {
      childProps.parentMatrixWorld = matrixWorld;
      childProps.multiplicationOrder = multiplicationOrder;
    }

    return React.cloneElement(child, childProps);
  };
}
