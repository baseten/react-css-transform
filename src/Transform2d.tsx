import PropTypes from 'prop-types';
import { mat2d, vec2 } from 'gl-matrix';

import { vec2Obj, vec2GlMatrix, mat2dGlMatrix } from './constants';
import { setVec2FromProp } from './utils';
import { useFactoryRef } from './useFactoryRef';
import { useRender } from './useRender';
import type {
  MultiplicationOrder,
  TransformChildren,
  Vec2Object,
} from './types';

const propTypes = {
  parentMatrixWorld: mat2dGlMatrix,
  multiplicationOrder: PropTypes.oneOf(['PRE', 'POST']).isRequired,
  translate: PropTypes.oneOfType([vec2GlMatrix, vec2Obj]),
  scale: PropTypes.oneOfType([vec2GlMatrix, vec2Obj, PropTypes.number]),
  rotate: PropTypes.number,
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
};

export type Transform2dProps = {
  children: TransformChildren<mat2d>;
  parentMatrixWorld?: mat2d;
  multiplicationOrder?: MultiplicationOrder;
  translate?: vec2 | Vec2Object;
  scale?: vec2 | Vec2Object | number;
  rotate?: number;
};

export type Apply2dTransformsParams = {
  parentMatrixWorld: mat2d;
  matrix: mat2d;
  matrixWorld: mat2d;
  vTranslation: vec2;
  vScale: vec2;
  multiplicationOrder: MultiplicationOrder;
  translate?: vec2 | Vec2Object;
  scale?: vec2 | Vec2Object | number;
  rotate?: number;
};

// TODO: somehow implement safe memoization with some sort of equality check?
// considering mutable objects
export function apply2dTransforms({
  parentMatrixWorld,
  matrix,
  matrixWorld,
  vTranslation,
  vScale,
  multiplicationOrder,
  translate,
  scale,
  rotate,
}: Apply2dTransformsParams) {
  const theta = typeof rotate === 'number' ? rotate : 0;

  mat2d.identity(matrix);

  setVec2FromProp(vTranslation, translate);
  setVec2FromProp(vScale, scale, 1);

  // T * R * S
  mat2d.translate(matrix, matrix, vTranslation);
  mat2d.rotate(matrix, matrix, theta);
  mat2d.scale(matrix, matrix, vScale);

  if (multiplicationOrder === 'PRE') {
    mat2d.multiply(matrixWorld, matrix, parentMatrixWorld);
  } else {
    mat2d.multiply(matrixWorld, parentMatrixWorld, matrix);
  }
}

const Transform2d = ({
  children,
  parentMatrixWorld,
  translate,
  scale,
  rotate,
  multiplicationOrder = 'POST',
}: Transform2dProps) => {
  const safeParentMatrixWorld = useFactoryRef<mat2d>(
    () => parentMatrixWorld || mat2d.create(),
  );
  const matrix = useFactoryRef<mat2d>(() => mat2d.create());
  const matrixWorld = useFactoryRef<mat2d>(() => mat2d.create());
  const vTranslation = useFactoryRef<vec2>(() => vec2.create());
  const vScale = useFactoryRef<vec2>(() => vec2.create());

  apply2dTransforms({
    parentMatrixWorld: safeParentMatrixWorld.current,
    matrix: matrix.current,
    matrixWorld: matrixWorld.current,
    vTranslation: vTranslation.current,
    vScale: vScale.current,
    multiplicationOrder,
    translate,
    scale,
    rotate,
  });

  const render = useRender<mat2d>({
    cssMatrixPrefix: 'matrix',
    matrixWorld,
    multiplicationOrder,
  });

  return render(children);
};

Transform2d.propTypes = propTypes;

export default Transform2d;
