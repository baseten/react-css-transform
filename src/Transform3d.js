import React from 'react';
import PropTypes from 'prop-types';
import { mat4, vec3 } from 'gl-matrix';

import { MULTIPLICATION_ORDER, vec3Shape, glMatrixType } from './constants';
import { setVec3FromProp } from './utils';

export default class Transform3d extends React.Component {
  static propTypes = {
    parentMatrixWorld: glMatrixType,
    multiplicationOrder: PropTypes.oneOf([MULTIPLICATION_ORDER.PRE, MULTIPLICATION_ORDER.POST]),
    translate: PropTypes.oneOfType([glMatrixType, vec3Shape]),
    scale: PropTypes.oneOfType([glMatrixType, vec3Shape, PropTypes.number]),
    rotate: PropTypes.number,
    rotateAxis: PropTypes.oneOfType([glMatrixType, vec3Shape]),
    children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
  };

  static defaultProps = {
    parentMatrixWorld: mat4.create(),
    multiplicationOrder: MULTIPLICATION_ORDER.POST,
    rotateAxis: {
      x: 0,
      y: 0,
      z: 1,
    },
  };

  matrix = mat4.create();
  matrixWorld = mat4.create();
  vTranslation = vec3.create();
  vScale = vec3.create();
  vRotationAxis = vec3.create();

  render() {
    const { children, parentMatrixWorld, multiplicationOrder, translate, rotate, rotateAxis, scale } = this.props;
    const { matrix, matrixWorld, vTranslation, vScale, vRotationAxis } = this;

    mat4.identity(matrix);

    setVec3FromProp(vTranslation, translate);
    setVec3FromProp(vScale, scale, 1, 1, 1);
    setVec3FromProp(vRotationAxis, rotateAxis, 0, 0, 1);

    // T * R * S
    mat4.translate(matrix, matrix, vTranslation);
    mat4.rotate(matrix, matrix, rotate || 0, vRotationAxis);
    mat4.scale(matrix, matrix, vScale);

    if (multiplicationOrder === MULTIPLICATION_ORDER.PRE) {
      mat4.multiply(matrixWorld, matrix, parentMatrixWorld);
    } else if (multiplicationOrder === MULTIPLICATION_ORDER.POST) {
      mat4.multiply(matrixWorld, parentMatrixWorld, matrix);
    }

    if (typeof children === 'function') {
      return children({
        parentMatrixWorld: matrixWorld,
        multiplicationOrder,
        style: {
          transform: matrixWorld.join(','),
        },
      });
    }

    return React.Children.map(children, this.renderChild);
  }

  renderChild = child => {
    const { multiplicationOrder } = this.props;
    const { matrixWorld } = this;

    const childStyle = child.props.style || {};
    const style = {
      ...childStyle,
      transform: `matrix3d(${matrixWorld.join(',')})`,
    };

    const childProps = {
      style,
    };

    if (typeof child.type !== 'string') {
      childProps.parentMatrixWorld = matrixWorld;
      childProps.multiplicationOrder = multiplicationOrder;
    }

    return React.cloneElement(child, childProps);
  };
}
