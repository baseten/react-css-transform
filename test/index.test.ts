import * as index from '../src/index';

describe('index', () => {
  test('export modules for react-css-transform', () => {
    expect(index).toMatchSnapshot();
  });
});
