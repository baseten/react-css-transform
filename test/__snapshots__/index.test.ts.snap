// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index export modules for react-css-transform 1`] = `
{
  "Transform2d": [Function],
  "Transform3d": [Function],
  "mat2dGlMatrix": [Function],
  "mat4GlMatrix": [Function],
  "mat4GlMatrixValidator": [Function],
  "use2dTransformations": [Function],
  "use3dTransformations": [Function],
  "useFactoryRef": [Function],
  "vec2GlMatrix": [Function],
  "vec2Obj": [Function],
  "vec3GlMatrix": [Function],
  "vec3Obj": [Function],
}
`;
